File: scratch/inject_clears.dart
import 'dart:io';

void main() {
  final dir = Directory('lib/providers');
  final files = dir.listSync().whereType<File>();

  for (final file in files) {
    if (!file.path.endsWith('_provider.dart') || file.path.endsWith('auth_provider.dart')) continue;
    
    var content = file.readAsStringSync();
    
    if (!content.contains('session_cleanup_coordinator.dart')) {
      content = content.replaceFirst("import 'package:flutter/material.dart';", "import 'package:flutter/material.dart';\nimport 'session_cleanup_coordinator.dart';");
      if (!content.contains('session_cleanup_coordinator.dart')) {
        content = "import 'session_cleanup_coordinator.dart';\n$content";
      }
    }
    
    final classNameMatch = RegExp(r'class\s+(\w+)\s+extends\s+ChangeNotifier\s*\{').firstMatch(content);
    if (classNameMatch == null) continue;
    
    final className = classNameMatch.group(1);
    final constructorRegex = RegExp(r'(' + className! + r'\s*\([^)]*\)\s*(?::\s*super\([^)]*\)\s*)?\{)');
    
    if (constructorRegex.hasMatch(content)) {
      content = content.replaceFirstMapped(constructorRegex, (match) {
        return '${match.group(1)}\n    SessionCleanupCoordinator.registerCleanupTask(clear);';
      });
    } else {
      content = content.replaceFirst(
        RegExp(r'(class\s+' + className + r'\s+extends\s+ChangeNotifier\s*\{)'),
        '\$1\n  $className() {\n    SessionCleanupCoordinator.registerCleanupTask(clear);\n  }\n'
      );
    }
    
    if (!content.contains('void clear()')) {
      final lastBrace = content.lastIndexOf('}');
      if (lastBrace != -1) {
        content = content.substring(0, lastBrace) + '\n  void clear() {\n    // Add specific clear logic here\n    notifyListeners();\n  }\n' + content.substring(lastBrace);
      }
    }
    
    file.writeAsStringSync(content);
  }
  print('Done.');
}

File: scratch/inject_clears.py
import os
import re

providers_dir = '/Users/shehrozali/cybersentinel/lib/providers'

for filename in os.listdir(providers_dir):
    if not filename.endswith('_provider.dart') or filename == 'auth_provider.dart':
        continue
    filepath = os.path.join(providers_dir, filename)
    with open(filepath, 'r') as f:
        content = f.read()

    # Find the class name
    class_match = re.search(r'class\s+(\w+)\s+extends\s+ChangeNotifier\s*\{', content)
    if not class_match:
        continue
    class_name = class_match.group(1)

    # Check if we already imported session_cleanup_coordinator
    if 'session_cleanup_coordinator.dart' not in content:
        # Add import at the top
        content = content.replace("import 'package:flutter/material.dart';", "import 'package:flutter/material.dart';\nimport 'session_cleanup_coordinator.dart';")
        if "import 'session_cleanup_coordinator.dart';" not in content:
            # fallback
            content = "import 'session_cleanup_coordinator.dart';\n" + content

    # Add constructor if not exists, or modify existing
    constructor_regex = r'(' + class_name + r'\s*\([^)]*\)\s*(?::\s*super\([^)]*\)\s*)?\{)'
    if re.search(constructor_regex, content):
        content = re.sub(
            constructor_regex,
            r'\1\n    SessionCleanupCoordinator.registerCleanupTask(clear);',
            content
        )
    else:
        # Inject constructor right after class declaration
        content = re.sub(
            r'(class\s+' + class_name + r'\s+extends\s+ChangeNotifier\s*\{)',
            r'\1\n  ' + class_name + r'() {\n    SessionCleanupCoordinator.registerCleanupTask(clear);\n  }\n',
            content
        )

    # Add clear method at the end of the class before the last '}'
    if 'void clear()' not in content:
        clear_method = """
  void clear() {
    // Add specific clear logic here
    notifyListeners();
  }
}"""
        # Find last '}'
        last_brace_index = content.rfind('}')
        if last_brace_index != -1:
            content = content[:last_brace_index] + clear_method + content[last_brace_index+1:]

    with open(filepath, 'w') as f:
        f.write(content)

print("Injected clear() and Coordinator hooks into providers.")
